compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryGenerator.kt

```diff
@@ -89,146 +89,147 @@ internal class ProvidesMethodFactoryGenerator : CodeGenerator {
     val moduleClass = clazz.asClassName()
 
     val content = FileSpec.builder(packageName, className)
-        .addType(
-            TypeSpec
-                .classBuilder(factoryClass)
-                .addAnvilAnnotation()
-                .addSuperinterface(Factory::class.asClassName().parameterizedBy(returnType))
-                .apply {
-                  if (!isObject || parameters.isNotEmpty()) {
-                    primaryConstructor(
-                        FunSpec.constructorBuilder()
+        .apply {
+          val canGenerateAnObject = isObject && parameters.isEmpty()
+          val classBuilder = if (canGenerateAnObject) {
+            TypeSpec.objectBuilder(factoryClass)
+          } else {
+            TypeSpec.classBuilder(factoryClass)
+          }
+
+          classBuilder.addAnvilAnnotation()
+              .addSuperinterface(Factory::class.asClassName().parameterizedBy(returnType))
+              .apply {
+                if (!canGenerateAnObject) {
+                  primaryConstructor(
+                      FunSpec.constructorBuilder()
+                          .apply {
+                            if (!isObject) {
+                              addParameter("module", moduleClass)
+                            }
+                            parameters.forEach { parameter ->
+                              addParameter(parameter.name, parameter.providerTypeName)
+                            }
+                          }
+                          .build()
+                  )
+
+                  if (!isObject) {
+                    addProperty(
+                        PropertySpec.builder("module", moduleClass)
+                            .initializer("module")
+                            .addModifiers(PRIVATE)
+                            .build()
+                    )
+                  }
+
+                  parameters.forEach { parameter ->
+                    addProperty(
+                        PropertySpec.builder(parameter.name, parameter.providerTypeName)
+                            .initializer(parameter.name)
+                            .addModifiers(PRIVATE)
+                            .build()
+                    )
+                  }
+                }
+              }
+              .addFunction(
+                  FunSpec.builder("get")
+                      .addModifiers(OVERRIDE)
+                      .returns(returnType)
+                      .apply {
+                        val argumentList = parameters.asArgumentList(
+                            asProvider = true,
+                            includeModule = !isObject
+                        )
+                        addStatement("return $functionName($argumentList)")
+                      }
+                      .build()
+              )
+              .apply {
+                val builder = if (canGenerateAnObject) this else TypeSpec.companionObjectBuilder()
```

Comment:
And here. If it can be an object, then we don't need an additional companion object.